docs/notebooks/recovering_from_errors.pct.py

```diff
@@ -0,0 +1,144 @@
+
+        return {OBJECTIVE: trieste.data.Dataset(x, branin(x))}
+
+
+observer = FaultyBranin()
+
+# %% [markdown]
+# ## Set up the problem
+# We'll use the same set up as before, except for the acquisition rule, where we'll use `TrustRegion`, which will better illustrate how to recover.
+
+# %%
+import gpflow
+
+search_space = trieste.space.Box(
+    tf.cast([0.0, 0.0], tf.float64), tf.cast([1.0, 1.0], tf.float64)
+)
+initial_data = observer(search_space.sample(5))
+
+variance = tf.math.reduce_variance(initial_data[OBJECTIVE].observations)
+kernel = gpflow.kernels.Matern52(variance, [0.2, 0.2]) + gpflow.kernels.White(1e-12)
+gpr = gpflow.models.GPR(
+    initial_data[OBJECTIVE].astuple(), kernel, noise_variance=1e-5
+)
+gpflow.set_trainable(gpr.likelihood, False)
+models = {OBJECTIVE: trieste.models.GaussianProcessRegression(gpr)}
+
+acquisition_rule = trieste.acquisition.rule.TrustRegion()
+
+# %% [markdown]
+# ## Run the optimization loop
+#
+# The optimizer returns an `OptimizationResult`. We can use that object's `astuple` method to get a 2-tuple of the final result and the history of the successful optimization steps.
+
+# %%
+bo = trieste.bayesian_optimizer.BayesianOptimizer(observer, search_space)
+
+result, history = bo.optimize(15, initial_data, models, acquisition_rule).astuple()
+
+# %% [markdown]
+# We can see from the logs that the optimization loop failed, and this can be sufficient to know what to do next if we're working in a notebook. However, sometimes our setup means we don't have access to the logs. We'll pretend from here that's the case.
+
+# %% [markdown]
+# ## Handling success
+#
+# We don't know if the optimization completed successfully or not, so we'll only try to access and plot the data if it was successful. We can find out if this was the case with the `is_ok` attribute on the `Result`. If it was successful, we know there is data in the `result`, which we can `unwrap` and view.
+
+# %%
+if result.is_ok:
+    data = result.unwrap().datasets[OBJECTIVE]
+    print("best observation: ", tf.reduce_min(data.observations))
+
+# %% [markdown]
+# ## Handling failure
+#
+# If on the other hand, the optimization didn't complete successfully, we can fix our observer, and try again. We can try again by using the data, model and acquisition state from the last successful step, which is the last element of the `history`.
+#
+# Note we can view any `result` by printing it. We'll do that here to see what exception was caught.
+
+# %%
+if result.is_err:
+    print("result: ", result)
+
+    observer.manual_fix()
+
+    result, new_history = bo.optimize(
+        15 - len(history),
+        history[-1].datasets,
+        history[-1].models,
+        acquisition_rule,
+        history[-1].acquisition_state
+    ).astuple()
+
+    history.extend(new_history)
+
+# %% [markdown]
+# We can repeat the above until the optimization process is complete, or we can put the whole process in a loop
+
+# %%
+while result.is_err:
```

Comment:
This requires a first manual call to `bo.optimize()`: maybe worth adding a comment on that?